Model/Indexer/AbstractIndexer.php

```diff
@@ -0,0 +1,198 @@
+use Magento\Indexer\Model\ProcessManager;
+use Nosto\Tagging\Util\Benchmark;
+use Magento\Framework\App\ObjectManager;
+use Nosto\Tagging\Model\Indexer\Dimensions\ModeSwitcherInterface;
+use Magento\Store\Model\Store;
+use Traversable;
+use ArrayIterator;
+use InvalidArgumentException;
+use UnexpectedValueException;
+
+abstract class AbstractIndexer implements DimensionalIndexerInterface, IndexerActionInterface, MviewActionInterface
+{
+    /** @var NostoHelperAccount */
+    private $nostoHelperAccount;
+
+    /** @var NostoHelperScope */
+    private $nostoHelperScope;
+
+    /** @var NostoLogger */
+    public $nostoLogger;
+
+    /** @var ProcessManager */
+    private $processManager;
+
+    /** @var DimensionProviderInterface */
+    private $dimensionProvider;
+
+    /**
+     * AbstractIndexer constructor.
+     * @param NostoHelperAccount $nostoHelperAccount
+     * @param NostoHelperScope $nostoHelperScope
+     * @param NostoLogger $nostoLogger
+     * @param DimensionProviderInterface $dimensionProvider
+     * @param ProcessManager|null $processManager
+     */
+    public function __construct(
+        NostoHelperAccount $nostoHelperAccount,
+        NostoHelperScope $nostoHelperScope,
+        NostoLogger $nostoLogger,
+        DimensionProviderInterface $dimensionProvider,
+        ProcessManager $processManager = null
+    ) {
+        $this->nostoHelperAccount = $nostoHelperAccount;
+        $this->nostoHelperScope = $nostoHelperScope;
+        $this->nostoLogger = $nostoLogger;
+        $this->dimensionProvider = $dimensionProvider;
+        $this->processManager = $processManager;
+    }
+
+    /**
+     * Get ModeSwitcher class to later get the indexer mode
+     *
+     * @return ModeSwitcherInterface
+     */
+    abstract public function getModeSwitcher(): ModeSwitcherInterface;
+
+    /**
+     * Implement logic of single store indexing
+     *
+     * @param Store $store
+     * @param array $ids
+     */
+    abstract public function doIndex(Store $store, array $ids = []);
+
+    /**
+     * @return string
+     */
+    abstract public function getIndexerId(): string ;
+
+    /**
+     * @param array $ids
+     * @suppress PhanTypeMismatchArgument
+     */
+    public function doWork(array $ids = [])
+    {
+        $userFunctions = [];
+
+        switch ($this->getModeSwitcher()->getMode()) {
+            case DimensionModeConfiguration::DIMENSION_NONE:
+                $userFunctions[] = function () use ($ids) {
+                    $storesWithNosto = $this->nostoHelperAccount->getStoresWithNosto();
+                    foreach ($storesWithNosto as $store) {
+                        $this->doIndex($store, $ids);
+                    }
+                };
+                break;
+            case DimensionModeConfiguration::DIMENSION_STORE:
+                /** @var Dimension[] $dimension  */
+                foreach ($this->dimensionProvider->getIterator() as $dimension) {
```

Comment:
Your code implies that the provider that is injected is ALWAYS the `StoreDimensionProvider `

Reply:
Yes, this is the only DimensionProvider we support. Others will break the logic we have implemented